rtsp.go

```diff
@@ -658,11 +669,34 @@ func newRTSPCamera(ctx context.Context, _ resource.Dependencies, conf resource.C
 	}
 	cancelCtx, cancel := context.WithCancel(context.Background())
 	reader := gostream.VideoReaderFunc(func(_ context.Context) (image.Image, func(), error) {
-		latest := rc.latestFrame.Load()
-		if latest == nil {
+		if videoCodec(rc.currentCodec.Load()) == MJPEG {
+			img := rc.latestMJPEGImage.Load()
+			if img == nil {
+				return nil, func() {}, errors.New("no mjpeg frame yet")
+			}
+			return *img, func() {}, nil
+		}
+
+		if rc.latestFrame.Load() == nil {
 			return nil, func() {}, errors.New("no frame yet")
 		}
-		return *latest, func() {}, nil
+		frame := rc.latestFrame.Load()
```

Comment:
This still has the same bug -- a reader can get the latest frame, but before acquiring `frame.mu`, the "decoder" can deref the image -- freeing it.

This `latestFrame.Load` needs to be atomic with the incrementRefs. The (simplified) problem here is that there's a mutex on the frame object, but the important layer here is actually the `rc *rtspCamera`. Moving frames into/out of the `latestFrame` must be "atomic" with the reference counting for those frames.

Reply:
Ok so IIUC, there are two concurrency problems

1. ref decrements need to be atomic with `put`-ing
2. Swapping out the `latestFrame` needs to be atomic with ref counting.

The per-frame mutex solves 1. We need something that does 2.

We can use an additional mutex maybe called `frameSwapMutex` owned by `rc` that makes 2 happen?

Reply:
Went ahead and added it. Let me know if you think this is sufficient or the wrong direction etc

Reply:
> Frame ref decrements need to be atomic with put-ing

That's true for when updating the `latestFrame`, but not for calling the "release" function. I claim at least -- definitely would like you to convince yourself that claim is correct. Or come up with a counter-example for the claim.

> We can use an additional mutex maybe called frameSwapMutex owned by rc that makes 2 happen (additionally we could un-atomic the latestFrame pointer)?

That's right. We can also remove the per frame mutex in favor of atomics. What you have right now is correct in that the mutex protects concurrent access to the `refCount` variable. But I'm going to recommend using atomics here, 2 reasons:
* From a readability perspective, the current usage of the mutex makes it seem like it's needed for something.
* From a correctness perspective, because the lock is:
  * held in conjunction with the `frameSwapMutex` and
  * is held for longer than necessary
  * There's a risk of introducing a deadlock down the road.

Reply:
> That's true for when updating the latestFrame, but not for calling the "release" function. I claim at least -- definitely would like you to convince yourself that claim is correct. Or come up with a counter-example for the claim.

Because we aren't swapping in frame?